chain/neutrino_test.go

```diff
@@ -0,0 +1,209 @@
+	// The returned channel is closed once shutdown is complete.
+	callStartStop := func() <-chan struct{} {
+		done := make(chan struct{})
+
+		go func() {
+			defer close(done)
+
+			err := nc.Start()
+			require.NoError(t, err)
+			nc.Stop()
+			nc.WaitForShutdown()
+		}()
+
+		return done
+	}
+
+	// For each wanted restart, execute callStartStop and wait until the
+	// call is done before continuing to the next execution.  Waiting for
+	// a read from done forces all executions of callStartStop to be done
+	// sequentially.
+	//
+	// The test fails if all of the wanted restarts cannot be completed
+	// sequentially before the timeout is reached.
+	for i := 0; i < wantRestarts; i++ {
+		select {
+		case <-timeout:
+			t.Fatal("timed out")
+		case <-callStartStop():
+		}
+	}
+}
+
+// TestNeutrinoClientNotifyReceived verifies that a call to NotifyReceived sets
+// the client into the scanning state and that subsequent calls while scanning
+// will call Update on the client's Rescanner.
+func TestNeutrinoClientNotifyReceived(t *testing.T) {
+	var (
+		timeout                 = time.After(1 * time.Second)
+		nc                      = newMockNeutrinoClient()
+		wantNotifyReceivedCalls = 50
+		wantUpdateCalls         = wantNotifyReceivedCalls - 1
+	)
+
+	// executeCalls calls NotifyReceived() synchronously n times without
+	// blocking the test and requires no error after each call.
+	executeCalls := func(n int) <-chan struct{} {
+		var (
+			addrs []btcutil.Address
+			done  = make(chan struct{})
+		)
+
+		go func() {
+			defer close(done)
+
+			for i := 0; i < n; i++ {
+				err := nc.NotifyReceived(addrs)
+				require.NoError(t, err)
+			}
+		}()
+
+		return done
+	}
+
+	// Wait for all calls to complete or test to time out.
+	select {
+	case <-timeout:
+		t.Fatal("timed out")
+	case <-executeCalls(wantNotifyReceivedCalls):
+		// Require that the expected number of calls to Update were made
+		// once done sending all NotifyReceived calls.
+		mockRescan := nc.rescan.(*mockRescanner)
+		gotUpdateCalls := mockRescan.updateArgs.Len()
+		require.Equal(t, wantUpdateCalls, gotUpdateCalls)
+	}
+}
+
+// TestNeutrinoClientNotifyReceivedRescan verifies concurrent calls to
+// NotifyBlocks, NotifyReceived and Rescan do not result in a data race
+// and that there is no panic on replacing the rescan goroutine single instance.
+func TestNeutrinoClientNotifyReceivedRescan(t *testing.T) {
+	var (
+		addrs        []btcutil.Address
+		nc           = newMockNeutrinoClient()
+		timeout      = time.After(1 * time.Second)
+		wantRoutines = 100
+		msgCh        = make(chan string, wantRoutines)
+		defaultMsg   = "successfully called %s"
+	)
+
+	// Define closures to wrap desired neutrino client method calls.
+
+	// callRescan calls the Rescan() method and asserts it completes
+	// with no errors. Rescan() is called with the hash of an empty header
+	// on each call.
+	var (
+		emptyHeader = &wire.BlockHeader{}
+		startHash   = emptyHeader.BlockHash()
+	)
```

Comment:
You'd be fine with just `startHash := new(wire.BlockHeader).BlockHash()` and no `emptyHeader` declared, but NBD.